proposals/053-cryptography-extension.md

```diff
@@ -0,0 +1,124 @@
+
+## Continued improvements to `botan`
+
+While `botan-bindings` and `botan-low` are effectively frozen, small changes may still occur. However, `botan` is likely to see significant churn as we continue to improve and respond to feedback.
+
+As `botan` is a very "fresh" library, it is expected to see some amount of churn in its API as it is developed and refined further. The initial goals of functionality are nominally complete, but there is still a great amount of ways to improve the safety and user experience.
+
+Expected improvements include:
+
+- Managed / limited lifespan objects with bracket / guaranteed cleanup
+- Auto-zeroing of sensitive memory
+- Implementing data types and instances for libraries such as `random` and `mtl`
+- Conditional compilation suport for optional modules
+- Improvements to the `Make / Remake` factory methods used to generate bindings
+
+## Botan C++ development
+
+`Botan C++` is being actively developed, so this project will need continued maintenance in order to achieve continued parity. Furthermore, several valuable features in `Botan C++` are not exposed by the C FFI, and require C / C++ development in order to expose them:
+
+- Stream ciphers
+- Complete X509 infrastructure implementation
+- Complete TLS server implementation
+
+One of the previous stretch goals was to develop the C bindings for X509 certificate architecture, and to [contribute that code back](https://github.com/randombit/botan/issues/3627) to `Botan C++`. Substantial progress was made, but further work is required in order to complete this. 
+
+Another feature missing from the Botan C FFI is Stream Ciphers, and like the X509 bindings, it will require effort to extend the Botan C++ library in order to expose it to Haskell.
+
+A third feature missing from the Botan C FFI is TLS, as Botan provides a full TLS server implementation, which will also require effort to expose to the FFI.
+
+## Development of cryptographic typeclasses and data families
+
+Due to the insufficiency of the ADT approach towards managing algorithms, cryptographic typeclasses and data families have been brought back into scope as a deliverable after being declared out-of-scope in the original proposal.
+
+While a first attempt has already been made to typeclassify the various cryptographic primitives, this interface will require extensive refinement and feedback over the coming months, and is likely to experience significant change before settling down.
+
+It is expected that these typeclasses will result in improved per-algorithm type safety, and better ergonomics through more consistent handling of cryptographic primitives such as keys, nonces, and ciphertexts.
+
+## Improvements to installation / CI / Unit tests
+
+One of the largest warts in using these libraries is the installation of the Botan C++ library. Package distributions are available for some users and operating systems via `pkg-config`, but other users currently require manual installation.
```

Comment:
I think at the moment the only OS which can install Botan via package manager is macOS (`brew install botan`), while everyone else has to build from sources.